### PR TITLE
Refactor assets slice

### PR DESCRIPTION
**Pull request by hyphenized**
Refactored assets slice to an object indexed by asset ids. State updates emitted from the indexing service are diffed and handled faster with this approach.

Builds on top of #3530 and #3526

## To Test
Switch from main to this branch
- [ ] Check Balances
- [ ] Check Overview
- [ ] Check Swaps
- [ ] Check Send page
- [ ] Check Activities

Latest build: [extension-builds-3553](https://github.com/tahowallet/extension/suites/14390460411/artifacts/811161494) (as of Tue, 18 Jul 2023 18:59:27 GMT).

### CODE REVIEW

File: background/redux-slices/assets.ts [hunk 1]
```diff
@@ -11,80 +11,77 @@ import {
...
-            duplicateIndexes.forEach((id) => {
-              // Update only the metadata for the duplicate
-              mappedAssets[newAsset.symbol][id] = {
-                ...mappedAssets[newAsset.symbol][id],
-                metadata: newAsset.metadata,
+          if (isSmartContractFungibleAsset(existing) && newAsset.metadata) {
+            existing.metadata ??= {}
+
+            // Update verified status, token lists or discovery txs for custom assets
+            if (!isBaselineTrustedAsset(existing)) {
```
Comment (kkosiorowska) [inline background/redux-slices/assets.ts:58]: `isBaselineTrustedAsset` checks if asset is a  base asset or asset from token list. This means that only custom assets should be updated. Why then do we allow updates to property `tokenLists`. Shouldn't custom assets have this property empty? 🤔

File: background/redux-slices/tests/assets.unit.test.ts [hunk 2]
```diff
@@ -12,33 +12,37 @@ import reducer, { assetsLoaded, SingleAssetState } from "../assets"
...
 const pricesState: PricesState = {
-  [getFullAssetID(assetWithPricePoint)]: {
+  [getFullAssetID(asset)]: {
     USD: pricePoint,
   },
 }
 
 describe("Reducers", () => {
   describe("assetsLoaded", () => {
     test("updates cached asset metadata", () => {
```
Comment (kkosiorowska) [inline background/redux-slices/tests/assets.unit.test.ts:27]: Non-blocking

I wonder if it wouldn't be good if the description was more specific about what tests do.

File: background/services/indexing/index.ts [hunk 3]
```diff
@@ -964,11 +976,17 @@ export default class IndexingService extends BaseService<Events> {
...
-      })
-    )
+      this.chainService.supportedNetworks.map(async (network) =>
+        this.cacheAssetsForNetwork(network)
+      )
+    ).then((cachedAssetsPromises) => {
+      const cachedAssets = cachedAssetsPromises
+        .filter(isFulfilledPromise)
+        .map(({ value }) => value)
+        .flat()
```
Comment (kkosiorowska) [inline background/services/indexing/index.ts:986]: Is it possible to use `flatMap` here?

File: background/tests/earn/assets.mock.ts [hunk 4]
```diff
@@ -3,22 +3,25 @@ import { ETHEREUM } from "../../constants"
...
+      },
+      {
+        name: "Uniswap",
+        symbol: "UNI",
+        decimals: 18,
+        homeNetwork: ETHEREUM,
+        contractAddress: "0x0",
+      },
+    ].map((asset) => [getFullAssetID(asset), asset])
+  )
```
Comment (kkosiorowska) [inline background/tests/earn/assets.mock.ts:24]: Non-blocking

I wonder why we don't just use `createSmartContractAsset` here.

**Review by kkosiorowska — COMMENTED** [review]
Not tested yet, some minor comments.

**Comment by hyphenized** [comment]
Closed until reprioritization

**Pull request closed without merging**